src/main/java/org/apache/sysml/runtime/controlprogram/paramserv/ParamservUtils.java

```diff
@@ -418,4 +429,19 @@ public static ListObject accrueGradients(ListObject accGradients, ListObject gra
 		});
 		return accGradients;
 	}
+
+	/**
+	 * Force to recompile all the instruction on CP mode
+	 * @param program program
+	 */
+	public static void recompileToCP(Program program) {
+		// Force platform to single node
+		// because some SP instructions will remain unchanged
+		// due to the reset of platform to Spark
+		// when recompiling the instructions
+		DMLScript.RUNTIME_PLATFORM oldRtPlatform = DMLScript.rtplatform;
+		DMLScript.rtplatform = DMLScript.RUNTIME_PLATFORM.SINGLE_NODE;
+		Recompiler.recompileProgramBlockHierarchy2Forced(program.getProgramBlocks(), 0, new HashSet<>(), LopProperties.ExecType.CP);
+		DMLScript.rtplatform = oldRtPlatform;
```

Comment:
This seems to be some outdated code. During the merge of a previous PR, I fixed the recompilation to CP which does not require modifying the global static flags (which could conflict with concurrent recompilation if paramserv is used inside parfor) - instead we can simply call `Recompiler.recompileProgramBlockHierarchy2Forced` with a given `ExecType et` to force all instructions into this `et` .

Reply:
In fact, I have a question about how to launch the script on spark mode. I found that we could execute the script by creating *MLContext* or by using *java SystemML.jar -exec spark*. So what is the difference here?

Reply:
To answer that I have to separate two things here (1) APIs and (2) execution modes, both of which are orthogonal except for certain APIs that only support a limited set of execution modes. 

Regarding APIs, we support command line, MLContext, JMLC, ML pipelines, and Keras2DML/Caffe2DML. Command lines itself covers several types: you can run it standalone through Java (as you show above), via the spark-submit script (where the SystemML's driver runs in Spark's driver process) and through the hadoop script (where SystemML's driver runs in a client process or YARN container).

However, you're actual question is more about execution modes, which you can influence with the command line flag -exec. There we support singlenode (all operations in CP), hadoop (all matrix operations in MR), spark (all matrix operations in SPARK), hybrid (CP or MR per operation), and hybrid_spark (CP or SPARK per operation). Note that hybrid_spark is set as the default if you come through MLContext or spark_submit command line, while singlenode is the default in JMLC. These three configurations of API/exec modes are what most applications use.